### PR TITLE
[Improvement]: Do not dispatch *_UPDATE events if there were no actual updates

### LINKED ISSUE #14607: [Improvement]: do not dispatch POST_UPDATE events if there were no actual updates
### Improvement description

If I do this:

```
$obj = DataObject::getById(123);
$obj->setFoo('aaa');

$obj->save(); // update
$obj->save(); // not an update
```

Currently, `DataObjectEvents::POST_UPDATE` (and probably assets, documents too?) will trigger **each time**, which creates a churn if you're using it to, for example, propagate events out of Pimcore via webhooks.

Ideally, if the object (asset, document) wasn't in some way updated (including calculated fields), the event wouldn't fire, meaning the save was basically as no-op, as if it didn't happen at all.

### PR DESCRIPTION
**Pull request by SamyMP**
## Changes in this pull request  
Resolves #14607

## Additional info  
This PR prevents the dispatch of `PRE_UPDATE` and `POST_UPDATE` events if no fields has been changed. 
Relies on PR #14401

### CODE REVIEW

**Comment by SamyMP** [comment]
It is technically ready to be reviewed, but I would like to have some approvals / know if it is ok to merge this improvement as it is

File: models/DataObject/Concrete/Dao.php [hunk 1]
```diff
@@ -272,10 +272,6 @@ public function update(bool $isUpdate = null): void
                     $data[$fieldName] = $insertData;
                     $this->model->set($fieldName, $fd->getDataFromResource($insertData, $this->model, $fieldDefinitionParams));
                 }
-
-                if ($this->model instanceof Model\Element\DirtyIndicatorInterface) {
```
Comment (lukmzig) [inline models/DataObject/Concrete/Dao.php:276]: Hi @SamyMP, can you please clarify for me why was this removed? :)

Comment (lukmzig) [reply models/DataObject/Concrete/Dao.php:276]: Hi @SamyMP, could you provide some information here? :)

**Comment by SamyMP** [comment]
@lukmzig I'll repost my message, just in case you can't see it:

I removed it because it's executed before the code reaches the POST_UPDATE listeners.

It means that if you modify a field (i.e. input), it should be marked as dirty on the POST_UPDATE listener, but this code removed the "dirty" property, which invalidate the rest of the modifications, since we need to know if the object was updated before dispatching the event

The stacktrace is :
-> Marking field as dirty
-> saving the object
-> PRE_UPDATE is dispatched
-> **dirty is removed from all fields**
-> POST_UPDATE is dispatched

That's why I removed the bold part in the list

**Review by lukmzig — CHANGES_REQUESTED** [review]
Hi @SamyMP, thank you for your reply and sorry, that I did not see your message sooner.
I think this is a valid and a good improvement, however we would like to ask you to change couple of things:

1. We would prefer if you could create 2 new events which would be triggered only if something was changed and keep POST_UPDATE and PRE_UPDATE with the original behavior?
2. Could you please also add a short comment/documentation for these 2 new events?
3. Could you please add these new events also for Assets and Documents? In that way it would be more consistent.

Thank you very much and let me know if you have any questions or thoughts about this.

**Comment by SamyMP** [comment]
Hi @lukmzig, thank you for the feedback. I'll try to add these improvement when I have time, I'll let you know when it's done

**Comment by SamyMP** [comment]
Hi @lukmzig, I started working on the suggestions you proposed, here's what I found : 

We could not add this implementation to the Assets and Documents because this event heavily relies on the `DirtyIndicatorTrait`. I tried to get the content of `dirtyFields` in case of an Asset or Document, but couldn't get anything else than "userModification". I checked in the code and it doesn't seem any Document or Asset's fields are mark as dirty while saving the elements.

Also, about the naming, I would have renamed the `PRE_UPDATE` and `POST_UPDATE` events `PRE_SAVE` and `POST_SAVE` with the current implementation, and implement this PR in the `UPDATE` events as it would have made more sense. But as you said, modifying this would introduce a big breaking change.

Currently, I don't know if I should proceed with this PR or just document the `DirtyIndicatorTrait` and encourage people to use it if they have a use case where they need to check what fields have changed. 
Also, we could add the modified fields as arguments of `PRE_UPDATE` and `POST_UPDATE` events by default so that it's easier to implement...

For this PR, I will commit my work on your first point, which was to create a 2nd separate event instead of modifying the original ones, and the part about adding a modified fields array as event argument

What do you think ?

**Comment by brusch** [comment]
@SamyMP Sorry, for the late answer. I agree with you, I don't think it makes sense to proceed with this approach/PR and we should rather focus on using `DirtyIndicatorTrait`. It just feels more consistent to me as well 😊 
Since the `DirtyIndicatorTrait` is there anyway, I also don't think it's necessary to pass the changed fields as an attribute of the event, since you can get the changed fields in the event from the trait directly. 

Can you maybe add a section to the docs, how to use it? 
Maybe this page here would be a good place for it - wdyt? 
https://github.com/pimcore/pimcore/blob/832a594ab8530598a1cb4a0e72d48d57e76f2b23/doc/05_Objects/03_Working_with_PHP_API.md#L1

**Pull request closed without merging**